components/suggestion/command_provider/app_command_parser.ts

```diff
@@ -231,229 +739,155 @@ export class AppCommandParser {
+            if (b.label.toLowerCase().startsWith(parsed.incomplete)) {
                 result.push({
-                    suggestion: base,
-                    complete: base,
-                    description: binding.description,
-                    hint: binding.hint || '',
+                    complete: b.label,
+                    suggestion: b.label,
+                    description: b.description,
+                    hint: b.hint || '',
                 });
             }
-        }
+        });
 
         return result;
     }
 
-    // getSuggestionsForCursorPosition computes subcommand/form suggestions
-    getSuggestionsForCursorPosition = async (pretext: string): Promise<AutocompleteSuggestion[]> => {
-        const binding = await this.getBindingWithForm(pretext);
-        if (!binding) {
-            return [];
-        }
-
-        let suggestions: AutocompleteSuggestion[] = [];
-        if (binding.bindings && binding.bindings.length) {
-            return this.getSuggestionsForSubCommands(pretext, binding);
-        }
-
-        const form = this.getFormFromBinding(binding);
-        if (form) {
-            const argSuggestions = await this.getSuggestionsForArguments(pretext, binding);
-            suggestions = suggestions.concat(argSuggestions);
-
-            // Add "Execute Current Command" suggestion
-            // TODO get full text from SuggestionBox
-            const fullText = pretext;
-            const missing = this.getMissingFields(fullText, binding);
-            if (missing.length === 0) {
-                const execute = this.getSuggestionForExecute(pretext);
-                suggestions = [execute, ...suggestions];
-            }
-        }
-
-        return suggestions;
-    }
-
-    // getMissingFields collects the required fields that were not supplied in a submission
-    getMissingFields = (fullText: string, binding: AppBinding): AppField[] => {
-        const form = this.getFormFromBinding(binding);
-        if (!form) {
-            return [];
-        }
-
-        const missing: AppField[] = [];
-
-        const values = this.getFormValues(fullText, binding);
-        for (const field of form.fields) {
-            if (field.is_required && !values[field.name]) {
-                missing.push(field);
+    // getParameterSuggestions computes suggestions for positional argument values, flag names, and flag argument values
+    getParameterSuggestions = async (parsed: ParsedCommand): Promise<AutocompleteSuggestion[]> => {
+        switch (Number(parsed.state)) {
+        case ParseState.StartParameter: {
+            // see if there's a matching positional field
+            const positional = parsed.form?.fields?.find((f: AppField) => f.position === parsed.position + 1);
+            if (positional) {
+                parsed.field = positional;
+                return this.getValueSuggestions(parsed);
             }
+            return this.getFlagNameSuggestions(parsed);
```

Comment:
Are positional and flag values exclusives? I mean, if I can I do something like: `/ci start-pipeline mypipeline --force=true` and `/ci start-pipeline --force=true mypipeline`, when I finish writing `/ci start-pipeline ` I should receive the suggestion for the positional argument "pipeline name" and the flag argument "force", right?

Reply:
they are supposed to be mixable in this parser, will add a couple more tests for it. It does create interesting problems, esp. with boolean flags (optional value).